src/util/subscription-set/expiring-sorted-set.ts

```diff
@@ -16,28 +19,83 @@ interface ExpiringSortedSetEntry<T> {
  * @typeParam T - the type of the set entries' values
  */
 export class ExpiringSortedSet<T> implements SubscriptionSet<T> {
-  map = new Map<string, ExpiringSortedSetEntry<T>>()
+  capacity: number
+  map: Map<string, LinkedListNode<ExpiringSortedSetEntry<T>>>
+  list: DoubleLinkedList
+
+  constructor(capacity: number) {
+    this.capacity = capacity
+    this.map = new Map<string, LinkedListNode<ExpiringSortedSetEntry<T>>>()
+    this.list = new DoubleLinkedList()
+  }
 
   add(key: string, value: T, ttl: number) {
-    this.map.set(key, {
-      value,
-      expirationTimestamp: Date.now() + ttl,
-    })
+    let node = this.map.get(key)
+    if (node) {
+      node.data = {
+        value,
+        expirationTimestamp: Date.now() + ttl,
+      }
+      this.moveToTail(node)
+    } else {
+      this.evictIfNeeded()
+      const data = {
+        value,
+        expirationTimestamp: Date.now() + ttl,
+      }
+      node = new LinkedListNode(key, data)
+      this.list.insertAtTail(node)
+      this.map.set(key, node)
+    }
   }
 
-  getAll(): T[] {
+  getAll(): PromiseOrValue<T[]> {
     const results: T[] = []
-    const now = Date.now()
+    for (const [key] of this.map.entries()) {
+      const value = this.get(key) as T
+      if (value) {
+        results.push(value)
+      }
+    }
+    return results
+  }
 
-    // Since we're iterating, might as well prune here
-    for (const [key, entry] of this.map.entries()) {
-      if (entry.expirationTimestamp < now) {
-        this.map.delete(key) // In theory, this shouldn't happen frequently for feeds
+  get(key: string): T | undefined {
+    const node = this.map.get(key)
+    if (node) {
+      const expired = node.data.expirationTimestamp <= Date.now()
+      if (expired) {
+        this.delete(key)
+        return undefined
       } else {
-        results.push(entry.value)
+        return node.data.value
       }
+    } else {
+      return undefined
     }
+  }
 
-    return results
+  delete(key: string): void {
+    const node = this.map.get(key)
+    if (node) {
+      this.list.remove(node)
+      this.map.delete(key)
+    }
+  }
+
+  private moveToTail(node: LinkedListNode) {
+    this.list.remove(node)
+    this.list.insertAtTail(node)
+  }
+
+  private evictIfNeeded() {
+    if (this.list.size >= this.capacity) {
+      logger.warn(`List reached maximum capacity, evicting least recently updated entry.`)
+      const node = this.list.removeHead()
+      if (node) {
+        logger.warn(`The node with key ${node.key} was removed`)
+        this.map.delete(node.key)
+      }
+    }
   }
```

Comment:
This is just me nitpicking now but could you combine these 2 logs into one?  So there's no confusion if something gets printed in between? Something like: 
```suggestion
  private evictIfNeeded() {
    if (this.list.size >= this.capacity) {
      const node = this.list.removeHead()
      if (node) {
        logger.warn(`List reached maximum capacity, evicting least recently updated entry. The subscription with key ${node.key} was removed.`)
        this.map.delete(node.key)
      }
    }
  }
```